File: tasks/legacy/delphi/DELPHI-01/workspace/legacy/FulfillmentStatus.pas
unit FulfillmentStatus;

interface

function StatusForStock(OnHand, Reserved, ReorderPoint: Integer): string;

implementation

uses
  SysUtils;

function StatusForStock(OnHand, Reserved, ReorderPoint: Integer): string;
var
  Available: Integer;
begin
  if (OnHand < 0) or (Reserved < 0) or (ReorderPoint < 0) or
     (Reserved > OnHand) then
    raise Exception.Create('invalid stock values');

  Available := OnHand - Reserved;
  if Available = 0 then
    Result := 'BACKORDER'
  else if Available <= ReorderPoint then
    Result := 'REORDER'
  else
    Result := 'READY';
end;

end.

File: tasks/legacy/delphi/DELPHI-01/workspace/src/main/java/FulfillmentStatus.java
public final class FulfillmentStatus {
    public enum Code { BACKORDER, REORDER, READY }

    private FulfillmentStatus() {}

    public static Code forStock(int onHand, int reserved, int reorderPoint) {
        if (onHand < 0 || reserved < 0 || reorderPoint < 0 || reserved > onHand) {
            throw new IllegalArgumentException("invalid stock values");
        }
        int available = onHand - reserved;
        if (available <= reorderPoint) return Code.REORDER;
        if (available == 0) return Code.BACKORDER;
        return Code.READY;
    }
}
